File: series1/arrange.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left, *right;
    TreeNode() : val(0), left(NULL), right(NULL) {}
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
    
};

void create(TreeNode* t, int& count, ifstream& file, bool left) {
    if (count == 0) return;
    int temp;
    file >> temp;
    if (temp == 0) {
        if (left) create(t, count, file, false);
        return;
    }
    count--;
    if (left) {
        t->left = new TreeNode(temp);
        create(t->left, count, file, true);
        create(t, count, file, false);
        return;
    }
    t->right = new TreeNode(temp);
    create(t->right, count, file, true);
};


TreeNode* temp;

int fix(TreeNode* t) {
    int l, r;
    if (t->left) l = fix(t->left);
    else l = t->val;
    if (t->right) r = fix(t->right);
    else r = t->val;
    if (l > r) {
        swap(t->left, t->right);
        return r;
    }
    return l;
}

string s = "";
void inorder(TreeNode* t) {
    if (!t) return;
    inorder(t->left);
    s += to_string(t->val) + " ";
    inorder(t->right);
}



int main(int argc, char* argv[]) {
    if (argc != 2) {
        cerr << "Error: Invalid number of arguments. Expected 2 arguments.\n";
        exit(EXIT_FAILURE);
    }

    string filename = argv[1];
    
    ifstream file(filename);

    if (!file.is_open()) {
        cerr << "Error: Unable to open file " << filename << "\n";
        return 1;
    }
    int count;
    file >> count;
    int t;
    file >> t;
    temp = new TreeNode(t);
    create(temp, count, file, true);
    t = fix(temp);
    inorder(temp);
    s.pop_back();
    cout << s;
    return 0;
}

File: series2/arrange.java
import java.io.File;
import java.util.Scanner;

public class arrange {
    static Scanner scanner;
    static int count;

    static class node {
        int val;
        node left, right;
        public node(int val) {
            this.val = val;
            this.left = null;
            this.right = null;
        }
    }

    public static void create(node t, boolean left) {
        if (count == 0) return;
        int temp = scanner.nextInt();
        if (temp == 0) {
            if (left) create(t, false);
            return;
        }

        count--;

        if (left) {
            t.left = new node(temp);
            create(t.left, true);
            create(t, false);
        } else {
            t.right = new node(temp);
            create(t.right, true);
        }
    }

    public static void swap(node n) {
        node temp = n.left;
        n.left = n.right;
        n.right = temp;
    }

    public static int fix(node t) {
        int l, r;
        if (t.left != null) l = fix(t.left);
        else l = t.val;
        if (t.right != null) r = fix(t.right);
        else r = t.val;
        if (l > r) {
            swap(t);
            return r;
        }
        return l;
    }

    static String s = "";

    public static void inorder(node t) {
        if (t == null) return;
        inorder(t.left);
        s = s + t.val + " ";
        inorder(t.right);
    }

    static node head;

    public static void main(String[] args) {
        String f = args[0];
        File file = new File(f);

        try (Scanner scanner = new Scanner(file)) {
            arrange.scanner = scanner;
            count = scanner.nextInt();
            head = new node(scanner.nextInt());
            count--;
            create(head, true);
            int t = fix(head);
            inorder(head);
            System.out.println(s);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
